platform/engineering/design-docs/2021-08-01-autonomous-deployments.md

```diff
@@ -0,0 +1,220 @@
+
+**Author(s):** Tim Wright  
+**Last Updated:** 08-06-2021  
+**Status:** **Draft** | In Review | Approved  
+**Approvers:** 
+
+- [ ] Mike Prichard 
+- [ ] Dror Matalon 
+- [ ] Michael Fleet 
+- [ ] Demian Ginther
+
+**Awareness and Review:**
+
+- [ ] Mike Chelen
+- [ ] Chris Valarida
+- [ ] General VFS Team Members
+
+## Table of Contents
+
+- [Autonomous Deployment Plan for vets-website React Applications](#autonomous-deployment-plan-for-vets-website-react-applications)
+  - [Table of Contents](#table-of-contents)
+  - [Overview](#overview)
+    - [Objective](#objective)
+    - [Desired Outcomes](#desired-outcomes)
+    - [Background](#background)
+    - [High Level Design](#high-level-design)
+      - [Production Deployments](#production-deployments)
+      - [Staging and Dev deployments](#staging-and-dev-deployments)
+  - [Specifics](#specifics)
+    - [Detailed Design](#detailed-design)
+      - [Roadmap](#roadmap)
+      - [Technologies Being Used](#technologies-being-used)
+    - [Work Estimates](#work-estimates)
+      - [Milestone: Update current system to function with new workflows](#milestone-update-current-system-to-function-with-new-workflows)
+      - [Milestone: Create and release staging/dev deployment behaviors](#milestone-create-and-release-stagingdev-deployment-behaviors)
+      - [Milestone: Create and release initial production behaviors with "allow" list](#milestone-create-and-release-initial-production-behaviors-with-allow-list)
+      - [Milestone: Create and release MVP production behaviors with "disallow" list](#milestone-create-and-release-mvp-production-behaviors-with-disallow-list)
+    - [Alternatives](#alternatives)
+    - [Post MVP](#post-mvp)
+    - [Questions](#questions)
+    - [Revision History](#revision-history)
+
+## Overview
+
+The document's purpose is to outline the minimum viable product (MVP) outcome to create autonomous deployment within the vets-website repository. The end of this document contains a post-MVP plan as well.
+
+### Objective
+
+VFS teams should be able to safely and reliably deploy their frontend applications to any environment at any time. This should happen without waiting for a scheduled deployment or manual intervention by Platform or other VFS teams. Deployments should also not cause significant risk to the build or to the site.
+
+While adding the ability to autonomously deploy, VA should still maintain the benefits of shared components (i.e. VFS teams are still using shared patterns/components for standardization and quality purposes, and we don’t see wide variance from one VA.gov experience to another).
+
+VFS teams should also be able to easily revert their change on their own, as needed.
+
+### Desired Outcomes
+
+- Increasing the production deployment rate
+- Increasing customer satisfaction
+- Decreasing support volume
+- Decreasing the average time from kickoff to production, by initiative complexity
+- Decreasing build failures
```

Comment:
What I'm gathering from this document is that the solution is to deploy to different environments using different branches. There's a lot of consideration about syncing between the branches and the transition from the daily deploy to this model, but I'm not picking up on how this three-branch approach is going to drive these goals forward.

- What are we doing to increase the deployment rate that we couldn't do with a single branch?
  - I see how we can deploy more quickly to `staging` and `dev`, since we won't run the tests.
  - What will we do differently in the `master` CI that allows us to deploy more quickly?
    - I see the section in "specific design" that says `build changed applications / run unit tests`. That should cut down on CI time, making more frequent deployments more feasible, but do we need a three-branch model to do this?
- I'd expect an increase in support volume due to the added complexity of the deployment.
  - "Do I merge to `master`, `staging`, or `dev`?"
  - "If I merge to a lower environment branch, do I have to make _another_ PR to merge from that branch into the others?"
  - Also the rebasing nightmares :see_no_evil: 
- I'd expect that increase in required domain knowledge and complexity to lead to a decrease in customer satisfaction.
- I don't see how this approach affects the average time from kickoff to production.
- Does "decreasing build failures" specifically refer to build failures blocking the production deploy? If so, how does this three-branch deploy model improve this?